# Copyright (C) 2015-2021, Wazuh Inc.
# Created by Wazuh, Inc. <info@wazuh.com>.
# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
"""SchemaValidator unit tests."""
import uuid
import random
from pathlib import Path
from unittest.mock import MagicMock, call, patch
import json
from ruamel.yaml import YAML
import pytest
from jsonschema.exceptions import ValidationError, UnknownType

from workflow_engine.schema_validator import SchemaValidator

@pytest.mark.parametrize('logger_mock',
                         [{'logger_to_patch':'workflow_engine.schema_validator.logger'}],
                         indirect=True)
def test_schema_validator_constructor(logger_mock: MagicMock):
    """Test SchemaValidator constructor normal flow.
    Check the state of the SchemaValidator instance variables after creation.

    Parameters
    ----------
    logger_mock : MagicMock
        logger fixture to check debug calls
    """
    schema_path = Path(__file__).parent.parent / 'schemas' / 'schema_v1.json'
    with open(schema_path, 'r') as schema_file:
        schema_data = json.load(schema_file)

    wf_file_path = Path(__file__).parent / 'data' / 'wf-ok.yaml'
    with open(wf_file_path, 'r') as file:
        yaml = YAML(typ='safe', pure=True)
        yaml_data = yaml.load(file)

    validator = SchemaValidator(schema_path, wf_file_path)
    assert validator.schema_data == schema_data
    assert validator.yaml_data == yaml_data
    calls = [call(f"Loading schema file: {schema_path}"), 
             call(f"Loading yaml file: {wf_file_path}")]
    logger_mock.debug.assert_has_calls(calls)


def test_schema_validator_constructor_ko():
    """"Test SchemaValidator constructor error flows.
    Check if the FileNotFoundError is raisen with a random file name.
    """
    schema_path = str(uuid.UUID(int=random.randint(0, 2^32)))
    with pytest.raises(FileNotFoundError, match=f'File "{schema_path}" not found.'):
        SchemaValidator(schema_path, schema_path)


def test_preprocess_data():
    """Test SchemaValidator preprocess_data.
    Check if the preprocess_data method does not raise exceptions with a valid file.
    """
    schema_path = Path(__file__).parent.parent / 'schemas' / 'schema_v1.json'
    wf_file_path = Path(__file__).parent / 'data' / 'wf-ok.yaml'
    validator = SchemaValidator(schema_path, wf_file_path)
    validator.preprocess_data()


@pytest.mark.parametrize('workflow_file, error_msg',
                         [('wf-ko-no-path-on-do.yaml',
                           "Missing required properties in 'with' for task: {'task': 'run-agent-tests-{agent}'"),
                          ('wf-ko-no-path-on-cleanup.yaml',
                           "Missing required properties in 'with' for task: {'task': 'allocate-manager'"),])
def test_preprocess_data_ko(workflow_file: str, error_msg: str):
    """Test SchemaValidator preprocess_data error flow.
    Check the ValidationError generated by invalid yml files.

    Parameters
    ----------
    workflow_file : str
        workflow yml file name.
    error_msg : str
        Error message to check
    """
    schema_path = Path(__file__).parent.parent / 'schemas' / 'schema_v1.json'
    wf_file_path = Path(__file__).parent / 'data' / workflow_file
    validator = SchemaValidator(schema_path, wf_file_path)
    with pytest.raises(ValidationError, match=error_msg):
        validator.preprocess_data()


def test_validate_schema():
    """Test SchemaValidator validate_schema with a valid yml file."""
    schema_path = Path(__file__).parent.parent / 'schemas' / 'schema_v1.json'
    wf_file_path = Path(__file__).parent / 'data' / 'wf-ok.yaml'
    validator = SchemaValidator(schema_path, wf_file_path)
    validator.validateSchema()


@pytest.mark.parametrize('logger_mock',
                         [{'logger_to_patch':'workflow_engine.schema_validator.logger'}],
                         indirect=True)
def test_validate_schema_ko(logger_mock: MagicMock):
    """Test SchemaValidator validate_schema error flows.
    Check the messages sent to the log when an invalid workflow yml file is used.

    Parameters
    ----------
    logger_mock : MagicMock
        logger fixture defined in conftest.py
    """
    schema_path = Path(__file__).parent.parent / 'schemas' / 'schema_v1.json'
    wf_file_path = Path(__file__).parent / 'data' / 'wf-ko-schema-error.yaml'
    validator = SchemaValidator(schema_path, wf_file_path)
    validator.validateSchema()
    logger_mock.error.assert_called_once()
    assert 'Schema validation error:' in logger_mock.error.call_args[0][0]

    logger_mock.error.reset_mock()
    validator = SchemaValidator(schema_path, wf_file_path)
    with patch('workflow_engine.schema_validator.jsonschema.validate', side_effect=UnknownType):
        validator.validateSchema()
    logger_mock.error.assert_called_once()
    assert 'Unexpected error at schema validation:' in logger_mock.error.call_args[0][0]
